File: c/stack/basic-calculator.c
int get_value(int left, int right, int op)
{
    if(op == '-')
        return left - right;
    else if(op == '+')
        return left + right;

    return 0;
}

int calculate(char* s) {
    int *op_st;
    int *val_st;
    int op_top;
    int val_top;
    int left,right;
    int len;
    int tmp;
    int op;

    len = strlen(s);

    op_st = malloc(sizeof(int) * len);
    val_st = malloc(sizeof(int) * len);
    op_top = val_top = -1;

    while(*s){
        if(*s == ' '){
            s++;
            continue;
        }

        if(isdigit(*s)){
            tmp = 0;
            while(isdigit(*s)){
                tmp = 10 * tmp + *s - '0';
                s++;
            }
            val_st[++val_top] = tmp;
            continue;
        }

        switch(*s){
            case '+':
            case '-':
                if(op_top > -1  && (op_st[op_top] == '+' ||
                    op_st[op_top] == '-')){
                    right = val_st[val_top--];
                    left = val_st[val_top--];
                    tmp = get_value(left, right, op_st[op_top--]);
                    val_st[++val_top] = tmp;
                }

                op_st[++op_top] = *s;
                break;
            case '(':
                op_st[++op_top] = *s;
                break;
            case ')':
                while(op_top > -1 && ((op = op_st[op_top--]) != '(')){
                    right = val_st[val_top--];
                    left = val_st[val_top--];
                    tmp = get_value(left, right, op);
                    val_st[++val_top] = tmp;
                }
                break;
            default:
                break;
        }
        s++;
    }

    while(op_top > -1){
        right = val_st[val_top--];
        left = val_st[val_top--];
        tmp = get_value(left, right, op_st[op_top--]);
        val_st[++val_top] = tmp;
    }

    tmp = val_st[val_top];
    free(op_st);
    free(val_st);

    return tmp;
}

File: go/stack/basic-calculator.go
package stack

import (
	"strconv"
)

func isDigit(val int32) bool {
	if val >= '0' && val <= '9' {
		return true
	}

	return false
}
var opW = map[int32]int {
	'(': 0,
	'-': 1,
	'+': 1,
	'*': 2,
	'/':2,
}

func getResult(op1, op2 int, op int32)(val int) {
	switch op {
	case '-':
		val = op1 - op2
	case '+':
		val = op1 + op2
	case '*':
		val = op1 * op2
	case '/':
		val = op1/op2
	}

	return val
}

func calculate(s string) int {
	st := make([]int, len(s))
	top := -1

	opSt := make([]int32, len(s)/2)
	opTop := -1

	for i := 0; i < len(s); i++ {
		c := int32(s[i])
		if c == ' ' {
			continue
		}

		if isDigit(c) {
			j := i+1
			for j <len(s) {
				if !isDigit(int32(s[j])) {
					break
				}
				j++
			}
			tmp := s[i:j]
			val, _ := strconv.Atoi(tmp)
			top++
			st[top] = val
			i = j - 1
			continue
		}

		if c == '(' {
			opTop++
			opSt[opTop] = c
			continue
		}

		if c == ')' || (opTop > -1 && opW[opSt[opTop]] >= opW[c]) {
			for opTop > -1 {
				topC := opSt[opTop]
				if topC == '(' || opW[topC] < opW[c] {
					break
				}

				opTop--

				var result int
				if top >= 1 {
					op2 := st[top]
					top--
					op1 := st[top]
					top--

					result = getResult(op1, op2, topC)
				}else {
					result = st[top]
					top--
					if topC == '-' {
						result = -result
					}
				}


				top++
				st[top] = result
			}
		}

		if c == ')' {
			opTop--
			continue
		}



		opTop++
		opSt[opTop] = c
	}

	for opTop > -1 {
		topC := opSt[opTop]
		opTop--

		var result int
		if top >= 1 {
			op2 := st[top]
			top--
			op1 := st[top]
			top--

			result = getResult(op1, op2, topC)
		}else {
			result = st[top]
			top--
			if topC == '-' {
				result = -result
			}
		}

		top++
		st[top] = result
	}

	return st[0]
}
